File: playwright-csharp/tests/Steps/RestfulBooker/UI/BookingSteps.cs
using Microsoft.Playwright;
using NUnit.Framework;
using Reqnroll;
using TestTracks.Playwright.CSharp.Specs.Data.DataBuilders.RestfulBooker;
using TestTracks.Playwright.CSharp.Specs.Pages.RestfulBooker.Admin;
using TestTracks.Playwright.CSharp.Specs.Pages.RestfulBooker.Booking;
using TestTracks.Playwright.CSharp.Specs.Support;
using AuthApiClient = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Auth.AuthApi;
using AuthToken = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Auth.AuthToken;
using BookingApiClient = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Booking.BookingApi;
using RoomApiClient = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Room.RoomApi;

namespace TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.UI;

[Binding]
public sealed class BookingSteps
{
    private const string SiteTarget = "restfulBooker";
    private const string ApiTarget = "restfulBookerApi";
    private const string AdminCredential = "restfulBookerAdmin";

    private readonly ScenarioState _state;
    private BookingPage? _bookingPage;
    private RoomsPage? _roomsPage;
    private AuthToken? _authToken;
    private int? _currentRoomId;
    private string? _currentRoomName;
    private int? _createdBookingId;

    public BookingSteps(ScenarioState state)
    {
        _state = state;
    }

    [Given("an available room exists")]
    public async Task GivenAnAvailableRoomExists()
    {
        var data = _state.Data.Load<AvailableRoomUiTestData>("ui-booking", "scenarios.availableRoom");

        var authApi = await _state.UseApiAsync(
            ApiTarget,
            (settings, request, evidence) => new AuthApiClient(settings, request, evidence));
        var roomApi = await _state.UseApiAsync(
            ApiTarget,
            (settings, request, evidence) => new RoomApiClient(settings, request, evidence));

        var admin = _state.Targets.Credential(AdminCredential);
        _authToken = await authApi.LogIn(AuthPayloadBuilder.FromCredential(admin));

        var room = RoomDataBuilder
            .ForScenario(_state.ScenarioId, data.Room)
            .BuildPayload();
        var createdRoom = await roomApi.CreateRoom(room, _authToken.Token);

        _currentRoomId = createdRoom.RoomId;
        _currentRoomName = createdRoom.RoomName;
        _state.Cleanup.Register(
            $"Delete room {createdRoom.RoomId}",
            () => roomApi.DeleteRoom(createdRoom.RoomId, _authToken!.Token));
    }

    [When("the guest books the room")]
    public async Task WhenTheGuestBooksTheRoom()
    {
        var data = _state.Data.Load<GuestBookingUiTestData>("ui-booking", "scenarios.guestCanCreateBooking");
        var page = await _state.UsePageAsync(SiteTarget);

        var homePage = new HomePage(page, _state.Settings);
        _bookingPage = new BookingPage(page, _state.Settings);
        var bookingApi = await _state.UseApiAsync(
            ApiTarget,
            (settings, request, evidence) => new BookingApiClient(settings, request, evidence));

        var bookingDetails = BookingDataBuilder
            .ForScenario(_state.ScenarioId, data.Booking)
            .ForRoom(_currentRoomId!.Value)
            .BuildFormData();

        await homePage.OpenBookingForRoom(_currentRoomId.Value, bookingDetails.BookingDates);
        _createdBookingId = await _bookingPage.BookRoom(bookingDetails);

        _state.Cleanup.Register(
            $"Delete booking {_createdBookingId.Value}",
            () => bookingApi.DeleteBooking(_createdBookingId.Value, _authToken!.Token));
    }

    [Then("the booking is shown as confirmed")]
    public async Task ThenTheBookingIsShownAsConfirmed()
    {
        await Assertions.Expect(_bookingPage!.Confirmation).ToBeVisibleAsync(
            new LocatorAssertionsToBeVisibleOptions { Timeout = _state.Settings.ExpectTimeoutMs });

        Assert.That(_createdBookingId, Is.GreaterThan(0));
    }

    [When("an administrator views the rooms")]
    public async Task WhenAnAdministratorViewsTheRooms()
    {
        var admin = _state.Targets.Credential(AdminCredential);
        var page = await _state.UsePageAsync(SiteTarget);

        var adminLoginPage = new AdminLoginPage(page, _state.Settings);
        _roomsPage = new RoomsPage(page, _state.Settings);

        await adminLoginPage.LogIn(
            admin.Username,
            admin.Password);
    }

    [Then("the room is visible to the administrator")]
    public async Task ThenTheRoomIsVisibleToTheAdministrator()
    {
        await Assertions.Expect(_roomsPage!.RoomNamed(_currentRoomName!)).ToBeVisibleAsync(
            new LocatorAssertionsToBeVisibleOptions { Timeout = _state.Settings.ExpectTimeoutMs });
    }

}

public sealed record AvailableRoomUiTestData(RoomDataDefaults Room);

public sealed record GuestBookingUiTestData(BookingDataDefaults Booking);

File: playwright-java/src/test/java/testtracks/steps/restfulbooker/ui/BookingSteps.java
package testtracks.steps.restfulbooker.ui;

import static com.microsoft.playwright.assertions.PlaywrightAssertions.assertThat;

import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.junit.jupiter.api.Assertions;
import testtracks.data.builders.restfulbooker.AuthPayloadBuilder;
import testtracks.data.builders.restfulbooker.BookingDataBuilder;
import testtracks.data.builders.restfulbooker.BookingDataDefaults;
import testtracks.data.builders.restfulbooker.RoomDataBuilder;
import testtracks.data.builders.restfulbooker.RoomDataDefaults;
import testtracks.pages.restfulbooker.admin.AdminLoginPage;
import testtracks.pages.restfulbooker.admin.RoomsPage;
import testtracks.pages.restfulbooker.booking.BookingPage;
import testtracks.pages.restfulbooker.booking.HomePage;
import testtracks.steps.restfulbooker.api.auth.AuthApi;
import testtracks.steps.restfulbooker.api.auth.AuthToken;
import testtracks.steps.restfulbooker.api.booking.BookingApi;
import testtracks.steps.restfulbooker.api.room.RoomApi;
import testtracks.support.ScenarioState;

public final class BookingSteps {
  private static final String SITE_TARGET = "restfulBooker";
  private static final String API_TARGET = "restfulBookerApi";
  private static final String ADMIN_CREDENTIAL = "restfulBookerAdmin";

  private final ScenarioState state;
  private BookingPage bookingPage;
  private RoomsPage roomsPage;
  private AuthToken authToken;
  private Integer currentRoomId;
  private String currentRoomName;
  private Integer createdBookingId;

  public BookingSteps(ScenarioState state) {
    this.state = state;
  }

  @Given("an available room exists")
  public void anAvailableRoomExists() {
    var data =
        state.data().load("ui-booking", "scenarios.availableRoom", AvailableRoomUiTestData.class);
    var authApi = state.useApi(API_TARGET, AuthApi::new);
    var roomApi = state.useApi(API_TARGET, RoomApi::new);
    var admin = state.targets().credential(ADMIN_CREDENTIAL);

    authToken = authApi.logIn(AuthPayloadBuilder.fromCredential(admin));

    var room = RoomDataBuilder.forScenario(state.scenarioId(), data.room()).buildPayload();
    var createdRoom = roomApi.createRoom(room, authToken.token());

    currentRoomId = createdRoom.roomId();
    currentRoomName = createdRoom.roomName();
    state
        .cleanup()
        .register(
            "Delete room " + createdRoom.roomId(),
            () -> roomApi.deleteRoom(createdRoom.roomId(), authToken.token()));
  }

  @When("the guest books the room")
  public void theGuestBooksTheRoom() {
    var data =
        state
            .data()
            .load("ui-booking", "scenarios.guestCanCreateBooking", GuestBookingUiTestData.class);
    var page = state.usePage(SITE_TARGET);

    var homePage = new HomePage(page, state.settings());
    bookingPage = new BookingPage(page, state.settings());
    var bookingApi = state.useApi(API_TARGET, BookingApi::new);

    var bookingDetails =
        BookingDataBuilder.forScenario(state.scenarioId(), data.booking())
            .forRoom(currentRoomId)
            .buildFormData();

    homePage.openBookingForRoom(currentRoomId, bookingDetails.bookingDates());
    createdBookingId = bookingPage.bookRoom(bookingDetails);

    state
        .cleanup()
        .register(
            "Delete booking " + createdBookingId,
            () -> bookingApi.deleteBooking(createdBookingId, authToken.token()));
  }

  @Then("the booking is shown as confirmed")
  public void theBookingIsShownAsConfirmed() {
    assertThat(bookingPage.confirmation())
        .isVisible(
            new com.microsoft.playwright.assertions.LocatorAssertions.IsVisibleOptions()
                .setTimeout(state.settings().expectTimeoutMs()));
    Assertions.assertTrue(createdBookingId > 0);
  }

  @When("an administrator views the rooms")
  public void anAdministratorViewsTheRooms() {
    var admin = state.targets().credential(ADMIN_CREDENTIAL);
    var page = state.usePage(SITE_TARGET);

    var adminLoginPage = new AdminLoginPage(page, state.settings());
    roomsPage = new RoomsPage(page, state.settings());
    adminLoginPage.logIn(admin.username(), admin.password());
  }

  @Then("the room is visible to the administrator")
  public void theRoomIsVisibleToTheAdministrator() {
    assertThat(roomsPage.roomNamed(currentRoomName))
        .isVisible(
            new com.microsoft.playwright.assertions.LocatorAssertions.IsVisibleOptions()
                .setTimeout(state.settings().expectTimeoutMs()));
  }

  public record AvailableRoomUiTestData(RoomDataDefaults room) {}

  public record GuestBookingUiTestData(BookingDataDefaults booking) {}
}
